File: 2472-MaximumNumberofNon-overlappingPalindromeSubstrings/2472-MaximumNumberofNon-overlappingPalindromeSubstrings.cpp
// Last updated: 29/9/2025, 3:23:45 am
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    static const int MOD = 1000000007;
    static const int BASE = 27;
    static const int MAXN = 2005; // matches your 2001 in Python

    long long powers[MAXN];

    Solution() {
        powers[0] = 1;
        for (int i = 1; i < MAXN; i++) {
            powers[i] = (powers[i - 1] * BASE) % MOD;
        }
    }

    int getAlp(char c) {
        return c - 'a';
    }

    int maxPalindromes(string st, int k) {
        int n = st.size();

        // prefix hashes
        vector<long long> prefix(n + 1, 0);
        long long s = 0;
        for (int i = 0; i < n; i++) {
            s = (s * BASE + getAlp(st[i])) % MOD;
            prefix[i + 1] = s;
        }

        // suffix hashes
        vector<long long> suffix(n + 1, 0);
        s = 0;
        for (int i = n - 1; i >= 0; i--) {
            s = (s * BASE + getAlp(st[i])) % MOD;
            suffix[i] = s;
        }

        vector<int> dp(n + 1, 0);

        for (int index = 0; index <= n; index++) {
            if (index > 0) dp[index] = max(dp[index], dp[index - 1]);

            for (int i = index + k - 1; i < n; i++) {
                if (st[i] == st[index]) {
                    int window = i - index + 1;

                    long long left = (prefix[i + 1] - (prefix[index] * powers[window]) % MOD + MOD) % MOD;
                    long long right = (suffix[index] - (suffix[i + 1] * powers[window]) % MOD + MOD) % MOD;

                    if (left == right) {
                        dp[i + 1] = max(dp[i + 1], dp[index] + 1);
                    }
                }
            }
        }

        return dp[n];
    }
};

File: 2472-MaximumNumberofNon-overlappingPalindromeSubstrings/2472-MaximumNumberofNon-overlappingPalindromeSubstrings.py
# Last updated: 9/15/2026, 4:55:01 PM
1class Solution:
2    def maxPalindromes(self, s: str, min_length: int) -> int:
3        n = len(s)
4        if min_length == 1: return n
5
6        dp = [[False] * n for _ in range(n + 1)]
7        for k in range(2):
8            for i in range(n):
9                dp[k][i] = True
10
11        palindrome = [[] for _ in range(n)]
12        for k in range(2, n + 1):
13            for i in range(k-1, n):
14                dp[k][i] = s[i] == s[i-k+1] and dp[k-2][i-1]
15                if k >= min_length and dp[k][i]:
16                    palindrome[i-k+1].append(i)
17
18
19        dp_two = [0] * (n+1)
20        for left in range(n):
21            dp_two[left] = max(dp_two[left], dp_two[left - 1])
22            for right in palindrome[left]:
23                dp_two[right + 1] = max(dp_two[right + 1], dp_two[left] + 1)
24
25        return max(dp_two[-1], dp_two[-2])